src/web/common.rkt

```diff
@@ -67,44 +67,90 @@
                               ,@(for/list ([line extra])
                                   `(li ,line)))))))))
 
+(define (render-comparison #:title [title #f] name a b )
+  (render-large #:title title name a `(span ([class "unit"]) " → ") b))
+
 (define (render-large #:title [title #f] name . values)
   `(div ,name ": " (span ([class "number"]
                           ,@(if title `([title ,title]) '()))
                          ,@values)))
 
-(define (render-comparison #:title [title #f] name a b )
-  (render-large #:title title name a `(span ([class "unit"]) " → ") b))
+(define (preprocess->c-like preprocess #:assert-has-parentheses [assert-has-parentheses #t])
+  (match preprocess
+    [(list 'sort variables ...)
+     (format
+      (if assert-has-parentheses
+          "assert(~a);"
+          "assert ~a;")
+      (string-join
+       (for/list ([a (in-list variables)]
+                  [b (in-list (cdr variables))])
+         (format "~a < ~a" a b))
+       " && "))]))
+
+(define (preprocess->c preprocess)
+  (preprocess->c-like preprocess #:assert-has-parentheses #t))
+
+(define (preprocess->java preprocess)
+  (preprocess->c-like preprocess #:assert-has-parentheses #f))
```

Comment:
This is setting yourself up for trouble in the future. C and Java already differ in some random syntactic feature for sort—when you add abs, are you going to add more flags? No, of course not, that'll be crazy. Instead, why not extract the `(string-join ... " && ")` portion into a helper function, and then call that from `preprocess->c` and `preprocess->java`. You won't need flags or anything like that, and extending to more types of preprocessing will just involve writing whatever the most natural code is for each.

More generally, you should take note of the fact that you typically write code that is more general than you need. That's a bad thing, not a good one. Ones ability to predict which generalizations you will want in the future is limited, and writing a generic function where a specialized one will do limits how you solve problems in the future.

Reply:
This is great feedback, much appreciated!